contracts/badger-ren/strategies/CurveSwapStrategy.sol

```diff
@@ -0,0 +1,117 @@
+
+    // Only swapper is allowed to make swaps.
+    bytes32 public constant SWAPPER_ROLE = keccak256("SWAPPER_ROLE");
+    // The curve exchange contract is registered in the address provider with ID 2.
+    uint256 public constant CURVE_REGISTRY_EXCHANGE_ID = 2;
+
+    address public curveRegistryAddressProvider;
+
+    function initialize(address _admin, address _registry) public initializer {
+        __AccessControl_init();
+        __ReentrancyGuard_init();
+
+        require(_admin != address(0x0), "must set admin address");
+        _setupRole(DEFAULT_ADMIN_ROLE, _admin);
+
+        require(_registry != address(0x0), "must set registry address provider address");
+        curveRegistryAddressProvider = _registry;
+    }
+
+    function swapTokens(
+        address _from,
+        address _to,
+        uint256 _amount,
+        uint256 _slippage
+    ) external nonReentrant onlySwapper returns (uint256 amount) {
+        (address registry, address pool, uint256 exchangeAmount) = _estimateSwapAmount(_from, _to, _amount);
+
+        _slippage = uint256(1e4).sub(_slippage);
+        uint256 minAmount = _amount.mul(_slippage).div(1e4);
+        require(exchangeAmount > minAmount, "slippage too high");
+
+        _approveBalance(_from, registry, _amount);
+        amount = ICurveRegistryExchange(registry).exchange(
+            pool,
+            _from,
+            _to,
+            _amount,
+            minAmount,
+            // Swap strategy caller is the receiver of the swap.
+            msg.sender
+        );
+        require(amount > minAmount, "swapped amount less than min amount");
+
+        return amount;
+    }
+
+    // Anyone can estimate swap amount as this fn is stateless.
+    function estimateSwapAmount(
+        address _from,
+        address _to,
+        uint256 _amount
+    ) external nonReentrant returns (uint256) {
+        (address _, address __, uint256 amount) = _estimateSwapAmount(_from, _to, _amount);
+        return amount;
+    }
+
+    function _estimateSwapAmount(
+        address _from,
+        address _to,
+        uint256 _amount
+    )
+        internal
+        returns (
+            address registry,
+            address pool,
+            uint256 amount
+        )
+    {
+        // NB: Pulling the registry exchange address w/in the tx because
+        // it seemed like this could change over time according to the docs.
+        registry = ICurveRegistryAddressProvider(curveRegistryAddressProvider).get_address(CURVE_REGISTRY_EXCHANGE_ID);
+        (pool, amount) = ICurveRegistryExchange(registry).get_best_rate(_from, _to, _amount);
+        return (registry, pool, amount);
+    }
+
+    function _approveBalance(
+        address _token,
+        address _spender,
+        uint256 _amount
+    ) internal {
+        if (IERC20Upgradeable(_token).allowance(address(this), _spender) < _amount) {
+            // Approve max spend.
+            IERC20Upgradeable(_token).approve(_spender, (1 << 64) - 1);
```

Comment:
what is the rationale for approving max spend?

Reply:
Save on gas?